### ISSUE TITLE
unxpected 'return.type.incompatible' error for covariant override

### ISSUE THREAD

**Issue by cushon** [typetools/checker-framework#1625]
Is this a bug? Doesn't `RunnableFuture<? extends @Nullable Object>` implement `Future<? extends @Nullable Object>`?

```java
import java.util.concurrent.AbstractExecutorService;
import java.util.concurrent.RunnableFuture;

abstract class A extends AbstractExecutorService {
  public RunnableFuture<?> submit(Runnable task) {
    return null;
  }
}
```

```
./checker-framework-2.2.1/checker/bin/javac -processor Nullness ~/T.java
T.java:5: error: [override.return.invalid] Incompatible return type.
  public RunnableFuture<?> submit(Runnable task) {
                       ^
  Method
    @Initialized @NonNull RunnableFuture<? extends @Initialized @Nullable Object> submit(@Initialized @NonNull A this, @Initialized @NonNull Runnable p0) in A
  cannot override
    @Initialized @NonNull Future<? extends @Initialized @Nullable Object> submit(@Initialized @NonNull ExecutorService this, @Initialized @NonNull Runnable p0) in java.util.concurrent.ExecutorService
  found   : @Initialized @NonNull RunnableFuture<? extends @Initialized @Nullable Object>
  required: @Initialized @NonNull Future<? extends @Initialized @Nullable Object>
1 error
```

**Comment by wmdietl** [comment]
Running with `-AprintVerboseGenerics` shows the complete bounds for the wildcard:

```
i1625.java:5: error: [override.return.invalid] Incompatible return type.
	  public RunnableFuture<?> submit(Runnable task) {
	                       ^
  Method
    @Initialized @NonNull RunnableFuture<?[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]> submit(@Initialized @NonNull A this, @Initialized @NonNull Runnable p0) in A
  cannot override
    @Initialized @NonNull Future<?[ extends @Initialized @Nullable Object super @Initialized @Nullable Void]> submit(@Initialized @NonNull ExecutorService this, @Initialized @NonNull Runnable p0) in java.util.concurrent.ExecutorService
  found   : @Initialized @NonNull RunnableFuture<?[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]>
  required: @Initialized @NonNull Future<?[ extends @Initialized @Nullable Object super @Initialized @Nullable Void]>
i1625.java:6: error: [return.type.incompatible] incompatible types in return.
		      return null;
		             ^
  found   : @FBCBottom @Nullable  null
  required: @Initialized @NonNull RunnableFuture<?[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]>
2 errors
```

in the return type, note `RunnableFuture<?[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]>` vs. `Future<?[ extends @Initialized @Nullable Object super @Initialized @Nullable Void]>` where the lower bound is `@NonNull` vs. `@Nullable`.

Looking at the specification in
https://github.com/typetools/checker-framework/blob/master/checker/jdk/nullness/src/java/util/concurrent/ExecutorService.java#L266

we see
`Future<@Nullable ?> submit(Runnable task);`
which enforces that the type argument of the return type is `@Nullable`.

In the javadoc to `submit`, note the sentence `The Future's {@code get} method will return {@code null} upon <em>successful</em> completion.`
This seems to indicate that the only possible value returned by the future is `null`, so the current specification seems sensible.
Does this make sense?

I agree that the error message needs improvement. Issue #788 is tracking that.

**Comment by wmdietl** [comment]
@cushon Can we consider this issue closed? #788 remains open for the error message improvement.
